### DISCUSSION TITLE
Uncertain how to proceed with unit testing - documentation unclear

### DISCUSSION THREAD

**Discussion by psibean** [prisma/prisma#12057] [Q&A] [answered]
Hey, I have a full code available on the [linked repository and branch](https://github.com/psibean/express-auth-example/tree/feature/initial-registration/backend). (run `npm run test` from within the backend folder linked here)

I'm following the [prisma-examples/typescript/test-express](https://github.com/prisma/prisma-examples/tree/latest/typescript/testing-express) example but it still seems to be using the real database.

I've been following the Unit Testing section in the documentation as well as the examples and I can't get ANY of them to actually use a mocked db / Prisma client - the tests always seem to operate on the actual database.

Any assistance and guidance in the right direction with the linked code would be appreciated.

Note that, in the example, if I update the `prisma-test-environment` DB_URL variable to `DATABASE_URL` instead, it throws an error saying the URL has to start with `postgres://` because of the specified protocol and the protocol does not allow the use of an environment variable

Edit: The example seems kinda restrictedas it provides no alternative to sqlite, so what's the point if you're stuck to that provider for all environments? It baffles me why prisma doesn't support a test db datasource, usually I'm happy to have another db setup which is migrated, used for testing, then destroyed just for testing purposes, but I don't see an easy way to do this with Prisma.

Even a simple test mode that uses an in-memory db would be pretty convenient.

**If I use the singleton approach, it doesn't seem to make any difference - the real database is still used.**

**Comment by austincrim** [comment]
Hey @psibean 👋 ,

> I'm following the [prisma-examples/typescript/test-express](https://github.com/prisma/prisma-examples/tree/latest/typescript/testing-express) example but it still seems to be using the real database.

This example is specifically for *integration* tests, not unit tests. It is setup to actually do operations on a sqlite database. It doesn't look like we currently have a unit testing example in that repo, I'll chat with our team about that 👍.

> I've been following the Unit Testing section in the documentation...

- Have you been following [this guide](https://www.prisma.io/docs/guides/testing/unit-testing)? It should set you up with a mock `PrismaClient` that will not perform any DB operations while testing. I was unable to find that kind of code in the repo you linked.

Thanks for your feedback, we are always looking to improve our docs/examples!

**Reply by psibean** [reply]
> * Have you been following [this guide](https://www.prisma.io/docs/guides/testing/unit-testing)? It should set you up with a mock `PrismaClient` that will not perform any DB operations while testing. I was unable to find that kind of code in the repo you linked.

Hey, that link is the one I was referring too, I omitted the code because it wasn't working.

So, the dependency injection option on that page doesn't really work for the code in the repo without re-structuring it so that the mock client can be passed in to everything in place of the real client. There's nothing necessarily wrong with making that change, but I also don't really like the idea of then having to explicitly make the type of my client variable either the real client, or the mocked one.

With that, if I follow the instructions of the singleton approach and attempt to mock the export, I've pushed this code to the same repo on [another branch here called testing/mock-prisma](https://github.com/psibean/express-auth-example/tree/testing/mock-prisma).

Following the guide you linked to do this, running the tests still seems to use the existing database and not the mock one, in fact, the singleton file doesn't seem to get executed at all when running jest. I'm thinking that perhaps there needs to be some configuration?

**After writing the above, I realized something**: The singleton file that the guide mentions for you to create probably needs to be a setupFile. I had a console log in my file and it wasn't logging when running tests. Now it does.

The guide is missing the crucial piece of information that the singleton fil needs to be a `setupFile`, and better yet, it should be a `setupFilesAfterEnv`.

Now, it appears the mock is working, but now the tests fail (but they pass with a clean real db).

Is the prisma mock going to actually operate and behave as expected, or is it literally only going to return values that I've mocked to return?

For my testing, I'd like to test the REAL results of the operation, not mocked values. I was hoping the prisma mock would just run an in-memory database and perform the actual queries on that so I can test real results, not mock a query call to return a value and then test it - this defeats the whole point of the test. In my case, I'm hitting routes and functions and I want to ensure that the queries they run are returning the expected results and hitting the expected errors.

Edit: I've pushed the code to the linked branch above where the mock is now working, but the tests are failing because the mock is returning undefined instead of the created entry.

**Another edit: A USECASE**

Here's why the mocking in the guide doesn't really work.

Let's say I have a function which runs a query which only selects X fields.

If I mock that call to return a hard coded object, I know that it's going to return those fields - of course it will, I've hard coded the return value.

But what if I've accidentally added one field there I don't want and I want to check in my test its not included in the result? My hard coded mock return will of course return true - but this obfuscates that there is potentially an issue in the actual function being tested.

Or better yet, what if the fields I'm returning in a query in a function are conditional and I want to ensure the function is returning the appropriate fields in the right cases?

**Reply by austincrim** [reply]
Thanks for the info! That's all really helpful.

> The guide is missing the crucial piece of information that the singleton file needs to be a `setupFile`, and better yet, it should be a `setupFilesAfterEnv`.

We include this detail under `Prerequisites`, but it is behind the `Add Jest to your project` dropdown and easily missed for such an important part of the guide. I will bring this to our docs team to see how we can make that clearer.

> Is the prisma mock going to actually operate and behave as expected, or is it literally only going to return values that I've mocked to return?

Providing return values to your mocks is common practice as the mocked module is typically not the one being tested. If you want Prisma to operate as usual during your tests, you would not mock it, but connect it to a test database by providing a different connection string when testing. This is the workflow we demonstrate with our [integration test example](https://github.com/prisma/prisma-examples/tree/latest/typescript/testing-express) and the pattern could be adapted for providers other than sqlite.

I definitely understand the utility of having a "test mode" or in-memory option for using Prisma in tests. I actually found a couple of existing issues around this idea.

- https://github.com/prisma/prisma/issues/11129
- https://github.com/prisma/prisma/issues/732

Feel free to leave a +1 on them to help our engineering teams prioritize.

**Reply by psibean** [reply]
Hey, thanks again!

**I got the tests working, post is mostly for info** note that this is a TypeScript ESM setup (not commonjs) and tests are working!
[Working code at this linked branch](https://github.com/psibean/express-auth-example/tree/testing/initial-integration) (requires a TEST_DATABASE_URL environment variable).

Apologies if I've come off aggressive or hostile at all.

So I updated the integration test example, given I' using TypeScript and ESM, I was getting the following error:

```
 prisma/prisma-test-environment.ts:10:41 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node12', or 'nodenext'.
```

But it doesn't tell me where that option should be configured, because node and jest do not have a --module CLI option, and I have module set in package.json and in tsconfig.

However, updating the file to .js and simply using `__dirname` resulted in this error:

```                                                                                                                                                                                                
    ReferenceError: __dirname is not defined in ES module scope                                                                                                                                                   
    This file is being treated as an ES module because it has a '.js' file extension
```

However keeping the environment as typescript and using the default `__dirname`  seemed to do the trick. Tests are now working as I was hoping them too.

These last couple of issues were probably more jest related than prisma.

**Reply by jszklarz** [reply]
> The guide is missing the crucial piece of information that the singleton file needs to be a setupFile, and better yet, it should be a setupFilesAfterEnv.

Just wanted to add that I had the exact same issue going through the docs and wasted a few hours on it.

**Reply by janpio** [reply]
Happy for any issues over at prisma/docs or optimally a PR, otherwise this will probably not get fixed. Thanks.

**Reply by austincrim** [reply]
i created an issue for this the first time around.

[#2961](https://github.com/prisma/docs/issues/2961)

**Comment by psibean** [comment, marked as answer]
Following the other comment chain on this post, the eventual solution for me was to adjust the [prisma-examples/typescript/test-express](https://github.com/prisma/prisma-examples/tree/latest/typescript/testing-express) example to work with postgres. Then because I was using TypeScript and ESM (not commonjs) I had a few other oddities, but despite being a module, I had to use `__dirname` in the environment class file.

[Working code at this linked branch](https://github.com/psibean/express-auth-example/tree/testing/initial-integration)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
